designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+```go
+// injector is used to inject certain fields to file templates.
+type injector struct {
+    // config stores the project configuration.
+    config config.Config
+
+    // boilerplate is the copyright comment added at the top of scaffolded files.
+    boilerplate string
+
+    // resource contains the information of the API that is being scaffolded.
+    resource *resource.Resource
+}
+```
+
+One of the plugin hooks allows plugins to set up their own flags. Commong flags (like `--group`, `--version`, `--kind`) are defined by the plugin system itself, and the resource created with those values populated will be passed to the plugins. `create api` and `create webhook` commands would require the GVK being passed as well to the plugins. Currently, `InjectResource` is a required hook for `create api` and `create webhook` commands and the entire resource (which contains the group, version and kind) is being passed to the plugins for these commands.
+
+All the raw flags of the plugins would have to be passed down to the plugins through injector. The plugin specific flags would have to be specified in the project configuration by the user. And the current implmentation is such that, every subcommand can optionally implement the `InjectConfig` hook which passes the project configuration to the plugins and also enable the plugins to modify it if required.
+
+Along with flags, boilerplate with empty copyright or an apache license and in go-style comments is also being passed down to the plugins through injector. If other kind of comments were to be supported, this logic would have to be added (scaffold markers do support YAML and Go-style comments).
+
+Additionally, we would pass the entire CLI environment to the plugin, as anything executing in subshell environment would be able to read the environment variables even if it's being restricted in terms of what Kubebuilder passes to it.
+
+For displaying help text information of the plugins, `UpdateMetadata` hook is the one that allows to modify the description and examples for the `--help` message of plugins. Currently, there is no support for help implementaion of plugins so logic would have to be added to the `UpdateMetadata` hook to display plugin specific help.
+
+#### Handling plugin failures across the chain
+
+If any plugin in the chain fails, we will halt the chain execution and report errors as one plugin may be dependent on the success of another. We can use `stderr` for the plugin to write back warnings/errors in a structured format that Kubebuilder understands and return a process exit code 1. The successful return type with exit code 0 contains the list of files that were scaffolded and the file contents will be given back to Kubebuilder.
+
+In the case of plugin failures even if one plugin across the chain fails, all the files that were scaffolded already until that point will not be written to disk to prevent a half committed state.
```

Comment:
How will this be enforced? Again, I have a plugin written in python and one written in java in the chain. How will I enforce that nothing was written unless it's part of the specification.

Reply:
The idea here is to provide an interface for passing file data between plugins in the chain. If your plugin writes something outside of this interface, it won't be passed to downstream plugins which is undesirable from any perspective. There's nothing stopping a plugin from breaking this interface though. That should be noted here.

Reply:
@estroz what's the "interface"? Is it basically the list of files and their contents that's being passed around?

Reply:
Effectively, yes (and error information).